protocols/autonat/src/v2/client/handler/dial_back.rs

```diff
@@ -90,3 +88,50 @@ impl ConnectionHandler for Handler {
         }
     }
 }
+
+struct State {
+    stream: libp2p_swarm::Stream,
+    oneshot: Option<oneshot::Receiver<io::Result<()>>>,
+}
+
+#[derive(Debug)]
+pub struct IncomingNonce {
+    pub nonce: Nonce,
+    pub sender: oneshot::Sender<io::Result<()>>,
+}
+
+fn perform_dial_back(
+    stream: libp2p_swarm::Stream,
+) -> impl futures::Stream<Item = io::Result<IncomingNonce>> {
+    let state = State {
+        stream,
+        oneshot: None,
+    };
+    futures::stream::unfold(state, |mut state| async move {
+        if let Some(ref mut receiver) = state.oneshot {
+            if receiver.await.is_err() {
+                return Some((Err(io::Error::from(io::ErrorKind::Other)), state));
+            }
```

Comment:
Let's `match` here and actually forward the error that we receive from the behaviour.

Reply:
yeah good point